### ISSUE TITLE
Applying AC-Norm to Layer Normalization or Instance Normalization

### ISSUE THREAD

**Issue by siny1998** [EndoluminalSurgicalVision-IMR/ACNorm#1]
Hi, great work on this! I’d like to implement the proposed AC-Norm in my backbone network, which currently uses either Layer Normalization or Instance Normalization. Could you please take a look at my modified implementation below and let me know if it’s correct?

Layer Normalization:
class _AffineCoLayerNorm(Module):

    def __init__(self, num_features, tem=1, eps=1e-5, affine=True):
        super(_AffineCoLayerNorm, self).__init__()
        self.num_features = num_features
        self.eps = eps
        self.affine = affine
        self.tem = tem

        if self.affine:
            self.weight = Parameter(torch.Tensor(num_features))
            self.bias = Parameter(torch.Tensor(num_features))
            self.weight_target = Parameter(torch.Tensor(num_features))
            self.bias_target = Parameter(torch.Tensor(num_features))
        else:
            self.register_parameter('weight', None)
            self.register_parameter('bias', None)
            self.register_parameter('weight_target', None)
            self.register_parameter('bias_target', None)

        self.weight.requires_grad = False
        self.bias.requires_grad = False
        self.reset_parameters()

    def reset_parameters(self):
        if self.affine:
            self.weight.data.uniform_()
            self.bias.data.zero_()
            self.weight_target.data.uniform_()
            self.bias_target.data.zero_()

    def forward(self, input):
        y_s = self.bias / (self.weight.abs() + self.eps)
        y_t = self.bias_target.data / (self.weight_target.data.abs() + self.eps)

        y_s = y_s.view(-1, 1)
        y_t = y_t.view(-1, 1)

        corr = -torch.abs(y_t - y_s.t())
        corr = torch.softmax(corr / self.tem, dim=1)

        sparse_corr = torch.where((corr - torch.diag(corr).unsqueeze(1)) >= 0, corr, torch.zeros_like(corr))

        trans_weight = sparse_corr.detach() @ self.weight_target.view(self.num_features, 1)
        trans_bias = sparse_corr.detach() @ self.bias_target.view(self.num_features, 1)
        trans_weight = trans_weight.squeeze()
        trans_bias = trans_bias.squeeze()

        z_trans = F.layer_norm(input, (self.num_features,), trans_weight, trans_bias, self.eps)
        z = F.layer_norm(input, (self.num_features,), self.weight_target, self.bias_target, self.eps)

        return z + z_trans
 
Instance Normalization:
class _AffineCoInstanceNorm(Module):

    def __init__(self, num_features, tem=1, eps=1e-5, affine=True):
        super(_AffineCoInstanceNorm, self).__init__()
        self.num_features = num_features
        self.eps = eps
        self.affine = affine
        self.tem = tem

        if self.affine:
            self.weight = Parameter(torch.Tensor(num_features))
            self.bias = Parameter(torch.Tensor(num_features))
            self.weight_target = Parameter(torch.Tensor(num_features))
            self.bias_target = Parameter(torch.Tensor(num_features))
        else:
            self.register_parameter('weight', None)
            self.register_parameter('bias', None)
            self.register_parameter('weight_target', None)
            self.register_parameter('bias_target', None)

        self.weight.requires_grad = False
        self.bias.requires_grad = False
        self.reset_parameters()

    def reset_parameters(self):
        if self.affine:
            self.weight.data.uniform_()
            self.bias.data.zero_()
            self.weight_target.data.uniform_()
            self.bias_target.data.zero_()

    def forward(self, input):
        y_s = self.bias / (self.weight.abs() + self.eps)
        y_t = self.bias_target.data / (self.weight_target.data.abs() + self.eps)

        y_s = y_s.view(-1, 1)
        y_t = y_t.view(-1, 1)

        corr = -torch.abs(y_t - y_s.t())
        corr = torch.softmax(corr / self.tem, dim=1)

        sparse_corr = torch.where((corr - torch.diag(corr).unsqueeze(1)) >= 0, corr, torch.zeros_like(corr))

        trans_weight = sparse_corr.detach() @ self.weight_target.view(self.num_features, 1)
        trans_bias = sparse_corr.detach() @ self.bias_target.view(self.num_features, 1)
        trans_weight = trans_weight.squeeze()
        trans_bias = trans_bias.squeeze()

        z_trans = F.instance_norm(input, weight=trans_weight, bias=trans_bias, eps=self.eps)
        z = F.instance_norm(input, weight=self.weight_target, bias=self.bias_target, eps=self.eps)

        return z + z_trans
